### PR TITLE
Use a combined frustum for culling when in VR

### LINKED ISSUE #13236: three.js VR renderer does frustum culling based on only the left eye, causing premature clipping on the right eye.
##### Description of the problem

The VR renderer does frustum culling based on the frustum of the left eye camera. It applies the same culling to the right eye. As a result, when an object leaves the left eye's view, the user can see the object flash out on the corner of the right eye periphery.

Saw  https://github.com/w3c/webvr/issues/203 was linked.

https://github.com/mrdoob/three.js/pull/11523

##### Three.js version

- [x] Dev
- [ ] r89
- [ ] ...

##### Browser

- [ ] All of them
- [ ] Chrome
- [x] Firefox
- [ ] Internet Explorer

##### OS

- [x] All of them
- [ ] Windows
- [ ] macOS
- [ ] Linux
- [ ] Android
- [ ] iOS

##### Hardware Requirements (graphics card, VR Device, ...)

VR Device

### PR DESCRIPTION
**Pull request by jsantell**
Fixes #13236, #13535, https://github.com/aframevr/aframe/issues/3488

There are a handful of ways of going about this. Ultimately though we need a view matrix and projection matrix to construct the frustum used in WebGLRenderer, and an ArrayCamera's matrices are currently only being used to generate that frustum (and sort order). This can be seen in the array camera example by moving `camera.position.x = 3`.

* Should this be a method on an ArrayCamera? What should its name be?
* If camera's exposed a Projection * ViewMatrix matrix, that could just be set as well
* Could also extend ArrayCamera for VRCamera with this feature
* Indifferent about adding an example, used mostly for testing
* There are some parts in the renderer that will use this new matrix world as the camera position (materials uCamPos) which we should probably continue using the reported matrix (between the eyes) from the system, in which case, `setProjectionFromUnion` maybe should only set the inverse world matrix, although the function isn't very intuitive.

Tested in polyfill, WebVR and WebXR (Daydream/Pixel 1)

Before, note the culling of cubes on the right eye when the cubes go out of view in the left:

frustum-old

After:

frustum-new

### CODE REVIEW

**Comment by fernandojsg** [comment]
Looking good Jordan! Although I don't feel like including this on ArrayCamera as is just a specific use case for array camera. We were discussing in the past about including a `getCullingVolume` for an arrayCamera (https://github.com/mrdoob/three.js/pull/11523), the generic use case would be just looping through all the frustum for each camera on the array, but we could end up implementing more optimal frustums like the one you exposed here.
I believe going that way would keep `ArrayCamera` generic enough to be used for other non-vr related stuff (CAVE, holograms,...) or even VR with more than 2 cameras (peripherical vision or so).

File: src/cameras/ArrayCamera.js [hunk 1]
```diff
@@ -16,7 +17,66 @@ ArrayCamera.prototype = Object.assign( Object.create( PerspectiveCamera.prototyp
 
 	constructor: ArrayCamera,
 
-	isArrayCamera: true
+	isArrayCamera: true,
+
+	/**
+	 * Assumes 2 cameras that are perpendicular and share an X-axis, and that
```
Comment (dmarcos) [inline src/cameras/ArrayCamera.js]: Did you mean parallel to each other or perpendicular to the shared x-axis?

**Comment by mrdoob** [comment]
@jsantell are you sure this also fixes #13535?

**Comment by jsantell** [comment]
I haven't tried explicitly with clipping planes for https://github.com/mrdoob/three.js/issues/13535 but should work

**Comment by Pluto8195** [comment]
@jsantell 
@mrdoob
Been using this commit for 2 weeks it does work for clipping planes!

**Comment by mrdoob** [comment]
Bad news.

This methods is producing a `projectionMatrix` with `NaN`s.

And, apparently, when culling a scene with a `projectionMatrix` with `NaN`s, what happens is that no objects get culled so it's essentially sending all the objects in the scene to the GPU regardless of where the user is looking at.

I've investigated a bit and so far I've found out that these lines:

https://github.com/mrdoob/three.js/blob/dev/src/renderers/webvr/WebVRUtils.js#L23-L24

Should be:

```js
var projL = cameraL.projectionMatrix.elements;
var projR = cameraR.projectionMatrix.elements;
```

But the resulting `projectionMatrix` still contains `NaN`s and also `Infinity`s.

@jsantell any chance you can take a look?

**Comment by mrdoob** [comment]
Now that the code works correctly we don't need `webvr_frustum` so I removed it.

This [fiddle](http://jsfiddle.net/1Lo48wcm/3/) proved to be more useful for testing the code anyway.

**Comment by Pluto8195** [comment]
was using before I saw this and noticed the projection seemed a bit warped, 
was this the cause?

> Bad news.
> 
> This methods is producing a `projectionMatrix` with `NaN`s.
> 
> And, apparently, when culling a scene with a `projectionMatrix` with `NaN`s, what happens is that no objects get culled so it's essentially sending all the objects in the scene to the GPU regardless of where the user is looking at.
> 
> I've investigated a bit and so far I've found out that these lines:
> 
> https://github.com/mrdoob/three.js/blob/dev/src/renderers/webvr/WebVRUtils.js#L23-L24
> 
> Should be:
> 
> ```js
> var projL = cameraL.projectionMatrix.elements;
> var projR = cameraR.projectionMatrix.elements;
> ```
> 
> But the resulting `projectionMatrix` still contains `NaN`s and also `Infinity`s.
>

**Comment by mrdoob** [comment]
> was using before I saw this and noticed the projection seemed a bit warped,
was this the cause?

We're not changing the projections matrices WebVR/WebXR is providing. We're only computing a new projectionMatrix to cull the objects in the scene.

**Pull request merged**